cgltf.h

```diff
@@ -2646,30 +2652,39 @@ cgltf_size cgltf_accessor_unpack_indices(const cgltf_accessor* accessor, cgltf_u
 	{
 		return 0;
 	}
+	if (index_component_size > out_component_size)
+	{
+		return 0;
+	}
 	const uint8_t* element = cgltf_buffer_view_data(accessor->buffer_view);
 	if (element == NULL)
 	{
 		return 0;
 	}
 	element += accessor->offset;
 
-	if (accessor->component_type == cgltf_component_type_r_32u && accessor->stride == sizeof(cgltf_uint))
+	if (index_component_size == out_component_size && accessor->stride == out_component_size)
 	{
-		memcpy(out, element, index_count * sizeof(cgltf_uint));
+		memcpy(out, element, index_count * index_component_size);
 	}
 	else
 	{
-		cgltf_uint* dest = out;
-
-		for (cgltf_size index = 0; index < index_count; index++, dest++, element += accessor->stride)
+		uint8_t* dest = (uint8_t*)out;
+		for (cgltf_size index = 0; index < index_count; index++, dest += out_component_size, element += accessor->stride)
 		{
-			*dest = (cgltf_uint)cgltf_component_read_index(element, accessor->component_type);
+			cgltf_size index_data = cgltf_component_read_index(element, accessor->component_type);
+			memcpy(dest, &index_data, index_component_size);
```

Comment:
One minor note is that this only works on little-endian platforms. Maybe that's fine though, as we have issues with endian conversion in general, but a comment would be nice here.

Reply:
Oh, this should also copy `out_component_size`, not `index_component_size`.

Reply:
Finally, I would need to measure this, but this might significantly regress the 16->32 bit conversion speed, because the memcpy here is very short and variable size so the compiler won't inline it. It might be better to explicitly handle index_component_size=1,2,4 with a switch and a fixed-length write directly.

Reply:
Sounds good. Do you want to profile it before we merge it?

Reply:
Yeah I'll need to double check perf. We never need to copy 1 byte indices here since even if out_component_size is 1, this forces us down the memcpy path, so I think this can just check if out_component_size is 2 and write a 2-byte or 4-byte index accordingly.

Reply:
> Oh, this should also copy `out_component_size`, not `index_component_size`.

I'm not sure this is correct. If the output array is `uint32_t` (`out_component_size`) but the each index is `uint16_t` (`index_component_size`), we want to:
1. Copy just the `uint16_t` of index data.
2. Move the index pointer to the next index (+ 16 bits).
3. Move the array pointer to the next index (+ 32 bits).

This is what the current loop is doing. Let me know if I'm misunderstanding your proposed change.

> I think this can just check if out_component_size is 2 and write a 2-byte or 4-byte index accordingly.

Keep in mind, this else block handles the case where `index_component_size < out_component_size`. According to the glTF specification, index data can be `UNSIGNED_BYTE`, `UNSIGNED_SHORT`, or `UNSIGNED_INT`. So the cases would be (as you said earlier):
* index size 1 to output size 2, 4, 8
* index size 2 to output size 4, 8
* index size 4 to output size 8

> variable size so the compiler won't inline it. It might be better to explicitly handle index_component_size=1,2,4 with a switch and a fixed-length write directly.

Yeah looking at Godbolt, I think you're right. Switching from the variable copy to the fixed-length one inlines on GCC and MSVC (didn't check Clang).
https://godbolt.org/z/6rqYYfvoq

I've done this in the latest change.

Reply:
At the cost of a little copy-paste, I moved the switch outside of the for loop so you don't pay the cost of the branch every loop.

Reply:
> Copy just the uint16_t of index data.

This will leave gaps in the target data: for every 4 bytes that the caller expects, you'll be only writing 2. This works without changes in gltfpack (which expects 32-bit output), but that's only because gltfpack uses zero-initialized memory for the output buffer which should not be expected.

Since we have separate loops anyway, I would recommend just writing uint16_t or uint32_t according to the *output* (critically!) component size. This will probably also mean we'll be able to drop the `case 1` as it becomes redundant, and this will actually remove the extra endianness concerns.

Reply:
Yeah I was thinking to require the user to provide a zero-initialized array. You're right that this isn't necessarily a reasonable expectation.

Ah okay I think I see what you're saying now. I've made this change and tested it for the 16-bit data -> 32-bit array case, works as expected.